src/ray/core_worker/experimental_mutable_object_manager.cc

```diff
@@ -159,230 +172,191 @@ void ExperimentalMutableObjectManager::DestroySemaphores(const ObjectID &object_
-
-  RAY_CHECK(static_cast<int64_t>(channel.mutable_object->header->data_size +
-                                 channel.mutable_object->header->metadata_size) <=
-            channel.mutable_object->allocated_size);
-  auto data_buf = SharedMemoryBuffer::Slice(
-      channel.mutable_object->buffer, 0, channel.mutable_object->header->data_size);
-  auto metadata_buf =
-      SharedMemoryBuffer::Slice(channel.mutable_object->buffer,
-                                channel.mutable_object->header->data_size,
-                                channel.mutable_object->header->metadata_size);
-
-  *result = std::make_shared<RayObject>(
+  int64_t version_read = 0;
+  RAY_RETURN_NOT_OK(channel->mutable_object->header->ReadAcquire(
+      sem, channel->next_version_to_read, &version_read));
+  RAY_CHECK_GT(version_read, 0);
+  channel->next_version_to_read = version_read;
+
+  size_t total_size = channel->mutable_object->header->data_size +
+                      channel->mutable_object->header->metadata_size;
+  RAY_CHECK_LE(static_cast<int64_t>(total_size), channel->mutable_object->allocated_size);
+  std::shared_ptr<SharedMemoryBuffer> data_buf =
+      SharedMemoryBuffer::Slice(channel->mutable_object->buffer,
+                                /*offset=*/0,
+                                /*size=*/channel->mutable_object->header->data_size);
+  std::shared_ptr<SharedMemoryBuffer> metadata_buf =
+      SharedMemoryBuffer::Slice(channel->mutable_object->buffer,
+                                /*offset=*/channel->mutable_object->header->data_size,
+                                /*size=*/channel->mutable_object->header->metadata_size);
+
+  result = std::make_shared<RayObject>(
       std::move(data_buf), std::move(metadata_buf), std::vector<rpc::ObjectReference>());
-
   return Status::OK();
 }
 
-Status ExperimentalMutableObjectManager::ReadRelease(const ObjectID &object_id)
+Status MutableObjectManager::ReadRelease(const ObjectID &object_id)
     ABSL_NO_THREAD_SAFETY_ANALYSIS {
-  auto reader_channel_entry = reader_channels_.find(object_id);
-  if (reader_channel_entry == reader_channels_.end()) {
+  ReaderChannel *channel = GetChannel<ReaderChannel>(reader_channels_, object_id);
+  if (!channel) {
     return Status::ObjectNotFound("Reader channel has not been registered");
   }
 
-  auto &channel = reader_channel_entry->second;
-  channel.lock->AssertHeld();
-
   PlasmaObjectHeader::Semaphores sem = GetSemaphores(object_id);
   RAY_RETURN_NOT_OK(
-      channel.mutable_object->header->ReadRelease(sem, channel.next_version_to_read));
+      channel->mutable_object->header->ReadRelease(sem, channel->next_version_to_read));
   // The next read needs to read at least this version.
-  channel.next_version_to_read++;
-  // This lock ensures that there is only one reader at a time. The lock is acquired in
-  // `ReadAcquire()`.
-  channel.lock->Unlock();
+  channel->next_version_to_read++;
 
+  channel->lock->Unlock();
```

Comment:
Same here, could you add back the comment?

Reply:
Done.